Lib/zipimport.py

```diff
@@ -461,6 +492,65 @@ def _read_directory(archive):
 
                 name = name.replace('/', path_sep)
                 path = _bootstrap_external._path_join(archive, name)
+
+                # Ordering matches unpacking below.
+                if (
+                    file_size == MAX_UINT32 or
+                    data_size == MAX_UINT32 or
+                    file_offset == MAX_UINT32
+                ):
+                    # need to decode extra_data looking for a zip64 extra (which might not
+                    # be present)
+                    while extra_data:
+                        if len(extra_data) < 4:
+                            raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
+                        tag = _unpack_uint16(extra_data[:2])
+                        size = _unpack_uint16(extra_data[2:4])
+                        if len(extra_data) < 4 + size:
+                            raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
+                        if tag == ZIP64_EXTRA_TAG:
+                            if (len(extra_data) - 4) % 8 != 0:
+                                raise ZipImportError(f"can't read header extra: {archive!r}", path=archive)
+                            values = [
+                                int.from_bytes(extra_data[i:i+8], 'little')
+                                for i in range(4, len(extra_data), 8)
+                            ]
+
+                            # N.b. Here be dragons: the ordering of these is different than
+                            # the header fields, and it's really easy to get it wrong since
+                            # naturally-occuring zips that use all 3 are >4GB and not
+                            # something that would be checked-in.
+                            # The tests include a binary-edited zip that uses zip64
+                            # (unnecessarily) for all three.
```

Comment:
Don't talk and what's in the test in a comment in the main sources. just end this sentence with "that use all 3 are >4GiB".

relatedly... I don't obviously see what you describe in a test?

Reply:
> relatedly... I don't obviously see what you describe in a test?

I think this refers to the extra tests in https://github.com/thatch/zipimport64/